consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+        let output_index = input.get_outpoint().get_output_index() as usize;
+        assert!(output_index <= tx.get_outputs().len());
+        Ok(tx.get_outputs()[output_index].get_value())
+    }
+
+    fn check_block_fee(&self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        let input_mlt = transactions
+            .iter()
+            .map(|x| {
+                x.get_inputs()
+                    .iter()
+                    .map(|input| {
+                        Self::get_input_value(&self.db_tx, input).expect("Couldn't get input")
+                    })
+                    .sum::<Amount>()
+            })
+            .sum();
+        let output_mlt: Amount = transactions
+            .iter()
+            .map(|x| x.get_outputs().iter().map(|output| output.get_value()).sum::<Amount>())
+            .sum();
+
+        // Check that fee is not negative
+        if output_mlt > input_mlt {
+            return Err(BlockError::Unknown);
+        }
+        Ok(())
+    }
+
+    fn check_tx_inputs(&self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        let mut total_value = Amount::new(0);
+        for tx in transactions {
+            for input in tx.get_inputs() {
+                let tx_index =
+                    Self::get_mainchain_index_by_outpoint(&self.db_tx, input.get_outpoint())?;
+                // If there is a wrong input index then it's cause a BlockError
+                let output = Self::get_output_by_outpoint(&self.db_tx, input.get_outpoint())?;
+
+                // Check is input has already spent
+                if tx_index
+                    .get_spent_state(input.get_outpoint().get_output_index())
+                    .map_err(BlockError::from)?
+                    == OutputSpentState::Unspent
+                {
+                    return Err(BlockError::Unknown);
+                }
+                // Check overflow
+                total_value = (total_value + output.get_value()).ok_or(BlockError::Unknown)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn check_tx_outputs(&self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        for tx in transactions {
+            for _output in tx.get_outputs() {
+                // TODO: Check tx outputs to prevent the overwriting of the transaction
+            }
+        }
+        Ok(())
+    }
+
+    // Connect new block
+    fn connect_tip(
+        &mut self,
+        new_tip_block_index: &BlockIndex,
+        best_block_id: &Option<Id<Block>>,
+    ) -> Result<(), BlockError> {
+        if &self.db_tx.get_best_block_id()? != new_tip_block_index.get_prev_block_id() {
+            return Err(BlockError::Unknown);
+        }
+        let block = self.get_block_from_index(new_tip_block_index)?.expect("Inconsistent DB");
+        let transactions = block.transactions();
+
+        if best_block_id.is_some() && !block.is_genesis(self.chain_config) {
```

Comment:
You don't need two checks. We have to distinguish between unrecoverable errors and invalid blocks. If best_block_id is not set and the block is not genesis, that's basically corruption in the database. Why is this being used here as if it's just a normal occurrence? This is wrong. Things that are irrecoverable should go to asserts. Then after the assert succeeds, you check for one reasonable condition.

Reply:
I left `is_genesis` check. Seems, it make more sense in readability